File: offer/56.go
package main

func singleNumber(nums []int) int {
	return _56_1(nums)
}

// perf: 89, 99
func _56_1(nums []int) (ans int) {
	for mask := 1; mask != 0; mask <<= 1 {
		cnt := 0
		for _, num := range nums {
			if num&mask != 0 {
				cnt++
			}
		}
		if cnt%3 == 1 {
			ans ^= mask
		}
	}
	return
}

File: offer/56.cpp
#include <unordered_set>
#include <vector>
using std::vector;

class Solution {
public:
	vector<int> singleNumbers(vector<int>& nums) {
		return _1(nums);
	}
private:
	/**
	 * @brief 
	 * @perf: [16, 100]
	 * @time:
	 * @space: O(n/2) = O(n)
	 */
	vector<int> _1(vector<int>& nums){
		std::unordered_set<int> us;
		for(int i = 0; i < nums.size(); ++i){
			auto itr = us.find(nums[i]);
			if(itr == us.end())
				us.insert(nums[i]);
			else
				us.erase(itr);
		}
		return vector<int> {*(us.begin()), *(++us.begin())};
	}
	vector<int> _2(vector<int>& nums){
		int ans = 0;
		for(auto num: nums)
			ans ^= num;
		return ans;
	}
};
